### PR TITLE
Persist full quest celebrations

### PR DESCRIPTION
**Pull request by KjellKod**
## Summary
Persist full Quest celebration markdown as durable journal artifacts so GitHub and the dashboard can navigate to the full celebration, not just the structured journal data.
- Adds a deterministic persistence helper for new completions and a one-off installer branch conflict celebration fixture.
- Keeps existing celebration_data JSON replay compatibility while linking journals and dashboard cards to full celebration markdown.

## Changes
- **Celebration persistence**:
  - Add `scripts/quest_celebrate/persist.py` to render and write full celebration markdown with metadata, fenced block-letter art, What Started This, cast, achievements, metrics, quote, quality tier, carry-over findings, and victory narrative.
  - Refuse to overwrite existing celebration files and surface a clear completion message instead.
- **Quest completion**:
  - Update `scripts/quest_complete.py` to write `docs/quest-journal/celebrations/<slug>_<date>.md` before creating new journal entries.
  - Add journal `Celebration` and `Full celebration` links only when the artifact exists.
- **Dashboard navigation**:
  - Add `JournalEntry.celebration_path` and loader support for journal Celebration links plus matching-file fallback.
  - Render a `Celebration: Open` link on quest cards without inlining the celebration body.
- **Documentation and fixtures**:
  - Add the one-off `installer-branch-conflict_2026-05-02` celebration artifact and link it from its journal.
  - Add this Quest's archived journal and celebration artifact, then rebuild `docs/dashboard/index.html`.
- **Tests**:
  - Cover completion persistence, no-overwrite behavior, required persisted sections, journal-backed `/celebrate` rendering, dashboard parsing/rendering, traversal rejection, and dashboard integration.

## Validation
- [x] Run `bash scripts/quest_validate-manifest.sh`. Expected result: both Quest manifest checks pass with no stale entries.
- [x] Run `python3 -m pytest tests/unit/test_quest_complete.py tests/unit/test_quest_celebrate.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q`. Expected result: all targeted tests pass.
- [x] **Walkthrough — journal to celebration**: Prerequisites: checkout this branch locally. Action: open `docs/quest-journal/installer-branch-conflict_2026-05-02.md` and follow the `Celebration` link. Expected: `docs/quest-journal/celebrations/installer-branch-conflict_2026-05-02.md` opens and preserves the title art in a fenced code block. Negative/edge case: existing journals without celebration artifacts should still render normally in dashboard tests.
- [ ] **Walkthrough — dashboard navigation**: Prerequisites: checkout this branch locally. Action: open `docs/dashboard/index.html` and find the Persist Celebrations or Installer Branch Conflict card. Expected: the card shows `Celebration: Open` and links to the matching markdown artifact under `docs/quest-journal/celebrations/`. Negative/edge case: stale or traversal-style Celebration paths are rejected by loader tests.
Watch for: the quality-score display for some journal-replayed celebrations can differ from the named quality tier; reviewers flagged that as pre-existing and out of scope.

## Notes
- This intentionally does not add a batch historical backfill system or overwrite policy beyond keeping existing files unchanged.

<pre>
     ▐▛███▜▌
    ▝▜█████▛▘
      ▘▘ ▝▝
Quest/Co-Authored by
Co-Authored-By: Claude Opus 4.6 &lt;noreply@anthropic.com&gt;
Co-Authored-By: Codex &lt;noreply@openai.com&gt;
in collaboration with KjellKod
</pre>

### CODE REVIEW

**Comment by KjellKod** [comment]
Addressed the two issues identified by cubic in commit 3e3cca9.

- Fixed the clipped persisted victory narrative in docs/quest-journal/celebrations/persist-celebrations_2026-05-03.md and updated the renderer to prefer full brief context over an already-truncated summary.
- Updated scripts/quest_dashboard/loaders.py so explicit Celebration links only produce dashboard links when the target file exists, matching the fallback behavior and avoiding stale Open links.

Validation rerun:
- bash scripts/quest_validate-manifest.sh
- python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q (211 passed)

File: scripts/quest_celebrate/persist.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    return CelebrationWriteResult(
+        path=target,
+        rel_path=rel_path,
+        created=True,
+        message=f"Celebration created: {target}",
+    )
+
+
+def _extract_labeled_paragraph(source: str, label: str) -> str:
+    pattern = rf"(?im)^\s*(?:[*_]*{label}[*_]*\s*:|[*_]*{label}[*_]*\s+-)\s*(.+)$"
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_celebrate/persist.py]: Good catch. Fixed in b574ee8 by teaching `_extract_labeled_paragraph()` to handle standalone `Problem:` / `Impact:` labels and consume the following paragraph. I also updated the persisted celebration artifact so `What Started This` contains the full problem and impact context.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (212 passed)

File: scripts/quest_dashboard/loaders.py [hunk 2]
```diff
@@ -229,6 +231,56 @@ def _parse_journal_entry(journal_path: Path, repo_root: Path) -> JournalEntry:
...
     )
 
+
+def _extract_celebration_path(
+    content: str,
+    journal_path: Path,
+    repo_root: Path,
+) -> Path | None:
+    """Extract and validate a repo-relative celebration path for a journal."""
+    journal_dir = journal_path.parent
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_dashboard/loaders.py:242]: Fixed in 34db486. The loader now validates explicit Celebration targets against `docs/quest-journal/celebrations/`, requires an existing regular `.md` file, and rejects links to other journal files such as `README.md`.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (213 passed)

File: docs/quest-journal/celebrations/persist-celebrations_2026-05-03.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+## Quest Quote
+
+> "Verdict: 🟡 Ready after small fixes"
+>
+> — Review finding
+
+## Victory Narrative
+
+**Problem:** Quest completion currently writes a structured journal and embedded `celebration_data`, but the full rich celebration markdown shown in chat is not persisted for most quests. Readers on GitHub and the dashboard can inspect dry journal data, but cannot revisit the block-letter title a... The quest finished with 1 plan iteration(s), 0 fix loop(s), and a persisted celebration artifact that future readers can open directly from the journal.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply docs/quest-journal/celebrations/persist-celebrations_2026-05-03.md]: Good catch. Fixed in 3e3cca9 and then improved in b574ee8. The persisted celebration artifact no longer contains the clipped `a...` narrative, and the renderer now prefers full brief context over an already-truncated summary when building the victory narrative.

Validation rerun after the latest related fixes:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (213 passed)

File: scripts/quest_dashboard/loaders.py [hunk 4]
```diff
@@ -229,6 +231,54 @@ def _parse_journal_entry(journal_path: Path, repo_root: Path) -> JournalEntry:
...
+
+    target = _celebration_link_target(content)
+    if target:
+        candidate = Path(target)
+        if candidate.is_absolute():
+            return None
+        resolved = (journal_dir / candidate).resolve()
+        try:
+            resolved.relative_to(allowed_root)
+            return resolved.relative_to(repo_root.resolve())
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_dashboard/loaders.py:257]: Good catch. Fixed in 3e3cca9 and tightened further in 34db486. The dashboard loader now only returns Celebration links for existing regular `.md` files under `docs/quest-journal/celebrations/`, so stale metadata and links to other journal files are rejected.

Validation rerun after the latest related fixes:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (213 passed)

File: .skills/celebrate/SKILL.md [hunk 5]
```diff
@@ -110,11 +110,17 @@ You have all the data from the artifacts. Now **create your own celebration**. B
 - Wrap the block-letter title art in a fenced code block (triple backticks).
 - Inside that code block, emit block-letter rows as plain text lines only.
 - Do **not** prefix block-letter rows with `#`, `-`, `>`, or any other markdown marker.
-- Keep the title art contiguous with no blank separator inserted inside the rows.
+- Use the full celebration block-letter style shown in the good example below (FIGlet/ANSI-Shadow-like `██╗`, `╚═╝`, `╔══` forms), not the compact 5-row fallback font when the artifact is meant to be read on GitHub.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/celebrate/SKILL.md:113]: Good catch. Fixed in 19182e2 by removing the stale compact-font `max ~5 letters per line` instruction and replacing it with the ANSI-style word-block guidance. The skill now consistently says to use readable FIGlet/ANSI-Shadow-like blocks for GitHub artifacts, splitting only over-wide words into adjacent chunks with no dropped letters.

Validation for the code change before this doc cleanup:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (217 passed)

File: scripts/quest_complete.py [hunk 6]
```diff
@@ -363,11 +381,27 @@ def main() -> int:
...
         journal_file = journal_dir / f"{slug}_{completion_date.isoformat()}.md"
         journal_rel_path = journal_file.relative_to(repo_root)
+        celebration_result = write_celebration_file(
+            journal_dir,
+            data,
+            completion_date,
+            journal_rel_path,
+        )
+        celebration_rel_path = (
+            celebration_result.rel_path if celebration_result.path.exists() else None
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_complete.py]: Fixed in 45a061c. The completion path now only exposes a journal Celebration link when the celebration artifact was created for that completion, or when an existing artifact has matching `quest-id` metadata. If the same slug/date path already exists for a different quest, the script keeps the file, omits the journal link, and returns no celebration path in the JSON payload.

Regression coverage added in `tests/unit/test_quest_complete.py` for the same-slug same-day collision case.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (218 passed)

File: scripts/quest_complete.py [hunk 7]
```diff
@@ -72,6 +73,31 @@ def _build_celebration_json(data: QuestData) -> dict:
...
+        return celebration_result.rel_path
+    return None
+
+
+def _celebration_file_matches_quest(path: Path, quest_id: str) -> bool:
+    if not quest_id:
+        return False
+    try:
+        text = path.read_text(encoding="utf-8")
+    except OSError:
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_complete.py]: Fixed in e0c51b9. Existing celebration metadata reads now catch both `OSError` and `UnicodeDecodeError`, so invalid UTF-8 is treated as a non-matching artifact instead of crashing quest completion.

Regression coverage added in `tests/unit/test_quest_complete.py`.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (220 passed)

File: scripts/quest_dashboard/loaders.py [hunk 8]
```diff
@@ -229,6 +231,58 @@ def _parse_journal_entry(journal_path: Path, repo_root: Path) -> JournalEntry:
...
+            return None
+        resolved = (journal_dir / candidate).resolve()
+        try:
+            resolved.relative_to(allowed_root)
+            if resolved.suffix != ".md" or not resolved.is_file():
+                return None
+            return resolved.relative_to(repo_root.resolve())
+        except ValueError:
+            return None
+
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_dashboard/loaders.py:260]: Fixed in e0c51b9. The dashboard fallback now uses the same quest-id guard as completion: it only links `celebrations/<journal-stem>.md` when that file is a regular Markdown artifact under `docs/quest-journal/celebrations/` and its `<!-- quest-id: ... -->` metadata matches the journal quest id. Mismatched same-slug artifacts are ignored.

Regression coverage added in `tests/unit/test_quest_dashboard_loaders.py`.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (220 passed)

File: scripts/quest_complete.py [hunk 9]
```diff
@@ -363,11 +406,30 @@ def main() -> int:
         journal_dir.mkdir(parents=True, exist_ok=True)
         journal_file = journal_dir / f"{slug}_{completion_date.isoformat()}.md"
         journal_rel_path = journal_file.relative_to(repo_root)
+        celebration_result = write_celebration_file(
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_complete.py]: Fixed in dafa1ad. Celebration persistence is now deferred until after the script has confirmed it will create a new journal entry. If the same slug/date journal already exists, completion skips writing the canonical celebration artifact as well, so it cannot occupy `celebrations/<slug>_<date>.md` for a skipped or different journal.

Regression coverage added in `tests/unit/test_quest_complete.py`.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (221 passed)

File: .skills/celebrate/SKILL.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (KjellKod) [reply .skills/celebrate/SKILL.md:113]: Fixed in d356190. The title-art guidance is now mirrored into `.agents/skills/celebrate/SKILL.md` and `.claude/skills/celebrate/SKILL.md` instead of only living under `.skills/celebrate/SKILL.md`, so all shipped skill entrypoints have the same yes/no examples and ANSI-shadow-like GitHub artifact guidance.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (222 passed)

File: docs/quest-journal/celebrations/persist-celebrations_2026-05-03.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+██████╗ ███████╗██████╗ ███████╗██╗███████╗████████╗
+██╔══██╗██╔════╝██╔══██╗██╔════╝██║██╔════╝╚══██╔══╝
+██████╔╝█████╗  ██████╔╝███████╗██║███████╗   ██║   
+██╔═══╝ ██╔══╝  ██╔══██╗╚════██║██║╚════██║   ██║   
+██║     ███████╗██║  ██║███████║██║███████║   ██║   
+╚═╝     ╚══════╝╚═╝  ╚═╝╚══════╝╚═╝╚══════╝   ╚═╝   
+
+ ██████╗███████╗██╗     ███████╗██████╗ 
+██╔════╝██╔════╝██║     ██╔════╝██╔══██╗
+██║     █████╗  ██║     █████╗  ██████╔╝██████╔╝
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply docs/quest-journal/celebrations/persist-celebrations_2026-05-03.md:20]: Fixed in d356190. The malformed persisted title row was regenerated from the final renderer, and I added `scripts/quest_celebrate/render_title_art.py` so future manual repairs use the same deterministic renderer as persistence instead of hand-editing rows.

Example:
`PYTHONPATH=scripts python3 scripts/quest_celebrate/render_title_art.py "Persist Celebrations" --max-width 70`

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (222 passed)

File: scripts/quest_celebrate/ascii_art.py [hunk 11]
```diff
@@ -282,6 +324,69 @@
...
+
+
+def _split_word_for_ansi_shadow(word: str, max_width: int) -> list[str]:
+    """Split very long words without truncating any characters."""
+    if _ansi_shadow_width(word) <= max_width:
+        return [word]
+
+    split_index = _balanced_ansi_shadow_split(word, max_width)
+    if split_index is None:
+        return [word]
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_celebrate/ascii_art.py:356]: Fixed in d356190. `_balanced_ansi_shadow_split()` now falls back to the largest prefix that fits when no two-way balanced split exists, then recurses on the remainder. That keeps very long single-word titles within `max_width` without dropping letters.

Regression coverage added for a word requiring more than two chunks.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/unit/test_quest_celebrate.py tests/unit/test_quest_complete.py tests/unit/test_quest_dashboard_loaders.py tests/unit/test_quest_dashboard_render.py tests/integration/test_build_quest_dashboard.py -q` (222 passed)

File: .claude/skills/celebrate/SKILL.md [hunk 12]
```diff
@@ -4,4 +4,266 @@ description: Play a quest completion celebration animation. Use when the user in
...
+- User asks to "celebrate", "play celebration", or "show the celebration" for a quest
+- Quest workflow reaches Step 7 (complete) and user chooses to celebrate
+- User points to a quest archive path or journal entry and asks to celebrate it
+
+## Process
+
+### Step 1: Resolve the Quest Source
+
+If the user provides an argument:
+1. If it's a full path (starts with `/` or `.`), use it directly
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .claude/skills/celebrate/SKILL.md]: Fixed in 9db6768. The path-resolution guidance now accepts absolute or relative paths only after resolving them under approved Quest artifact roots: `.quest/`, `.quest/archive/`, or `docs/quest-journal/`; otherwise the path should be rejected. I applied the wording across `.skills`, `.agents`, and `.claude` celebrate skill surfaces.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/ -q` (562 passed)

File: docs/quest-journal/celebrations/persist-celebrations_2026-05-03.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+██████╗ ███████╗██████╗ ███████╗██╗███████╗████████╗
+██╔══██╗██╔════╝██╔══██╗██╔════╝██║██╔════╝╚══██╔══╝
+██████╔╝█████╗  ██████╔╝███████╗██║███████╗   ██║   
+██╔═══╝ ██╔══╝  ██╔══██╗╚════██║██║╚════██║   ██║   
+██║     ███████╗██║  ██║███████║██║███████║   ██║   
+╚═╝     ╚══════╝╚═╝  ╚═╝╚══════╝╚═╝╚══════╝   ╚═╝   
+
+ ██████╗███████╗██╗     ███████╗██████╗ 
+██╔════╝██╔════╝██║     ██╔════╝██╔══██╗
+██║     █████╗  ██║     █████╗  ██████╔╝
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply docs/quest-journal/celebrations/persist-celebrations_2026-05-03.md]: Fixed in 9db6768. The persisted banner row is back in sync with the renderer output, preserving the full `CELEBR` chunk instead of a manually edited shorter row.

I also verified the direct CLI path now renders the same title art:
`python3 scripts/quest_celebrate/render_title_art.py "Persist Celebrations" --max-width 70`

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/ -q` (562 passed)

File: .agents/skills/celebrate/SKILL.md [hunk 14]
```diff
@@ -1,6 +1,270 @@
 ---
 name: celebrate
-description: Play a quest completion celebration animation. Use when the user invokes $celebrate or asks to celebrate a quest, or when a quest reaches the complete/archived state.
+description: Play a quest completion celebration animation. Use when the user invokes /celebrate, asks to celebrate a quest, or when a quest reaches the complete/archived state.
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply .agents/skills/celebrate/SKILL.md]: Fixed in 9db6768. The installed Codex skill surface now restores the `$celebrate` trigger wording in `.agents/skills/celebrate/SKILL.md`, while the slash-command wording remains on the Claude/project surfaces that use `/celebrate`.

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/ -q` (562 passed)

File: scripts/quest_celebrate/render_title_art.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python3
+"""Render persisted celebration title art."""
+
+from __future__ import annotations
+
+import argparse
+
+from quest_celebrate.ascii_art import ansi_shadow_title
```
[the remark this answers is not in the document]
Comment (KjellKod) [reply scripts/quest_celebrate/render_title_art.py]: Fixed in 9db6768. `render_title_art.py` now follows the direct-script import fallback pattern, so this works from a checkout without setting `PYTHONPATH`:

`python3 scripts/quest_celebrate/render_title_art.py "Persist Celebrations" --max-width 70`

Validation rerun:
- `bash scripts/quest_validate-manifest.sh`
- `python3 -m pytest tests/ -q` (562 passed)

File: .agents/skills/celebrate/SKILL.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (KjellKod) [reply .agents/skills/celebrate/SKILL.md]: /celebrate is what it looks like in Claude and $celebrate is what it looks like in codex. Both must work
